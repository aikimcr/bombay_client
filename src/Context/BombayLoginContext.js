import { createContext } from "react";

const BombayLoginContext = createContext({
  loggedIn: false,
  showLoginForm: false,
  setLoggedIn: null,
  setShowLogin: null,
});

export default BombayLoginContext;
