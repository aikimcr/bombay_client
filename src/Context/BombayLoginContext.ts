import { createContext, Dispatch } from "react";

interface BombayLoginContextValues {
  loggedIn: boolean;
  setLoggedIn?: Dispatch<boolean>;
}

export const BombayLoginContext = createContext({
  loggedIn: false,
  setLoggedIn: null,
});

export default BombayLoginContext;
