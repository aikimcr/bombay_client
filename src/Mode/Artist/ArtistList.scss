@import '../../global.scss';

.list-component {
    height: 100%;
    width: 100%;

    display: grid;
    grid-template-rows: $button-font-size + 10px auto;
    grid-template-areas:
        "header"
        "content"
}

.list-controls {
    grid-area: header;
    height: $button-font-size + 10px;
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 0 10px;
}

.artist-list-container {
    grid-area: content;
    @include scrollListContainer;
}