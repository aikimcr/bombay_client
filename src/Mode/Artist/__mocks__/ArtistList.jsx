import React from 'react';
import { ComponentMock } from '../../../testHelpers/mocks/ComponentMock';

export const ArtistList = (props) => {
  return (
    <ComponentMock
      {...props}
      name="ArtistList"
      className="mock-artist-list"
      testId="mock-artist-list"
    />
  );
};
