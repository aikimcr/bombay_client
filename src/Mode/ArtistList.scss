@import '../global.scss';

.artist-list-container {
    @include scrollListContainer;
}