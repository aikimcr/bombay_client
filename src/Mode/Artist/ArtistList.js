import { useState, useContext, createRef } from "react";

import "./ArtistList.scss";

import BombayLoginContext from "../../Context/BombayLoginContext";

import ArtistCollection from "../../Model/ArtistCollection";

import ArtistListItem from "./ArtistListItem";
import Artist from "./Artist";
import FormModal from "../../Modal/FormModal";
import useModelCollection from "../../Hooks/useModelCollection";

function ArtistList(props) {
  const topRef = createRef();

  const loginState = useContext(BombayLoginContext);

  const [artistCollection, refreshCollection] = useModelCollection({
    CollectionClass: ArtistCollection,
    topRef,
    loginState,
  });

  const [showAdd, setShowAdd] = useState(false);

  async function submitNewArtist(artistDef) {
    await artistCollection.current.save(artistDef);
    refreshCollection();
  }

  if (!loginState.loggedIn) return null;

  return (
    <div className="list-component">
      <div className="list-controls">
        <button className="btn" onClick={() => setShowAdd(true)}>
          New
        </button>
        <div className="title">Artists</div>
        <button className="btn" onClick={refreshCollection}>
          Refresh
        </button>
        <FormModal
          title="Add Artist"
          onClose={() => setShowAdd(false)}
          onSubmit={submitNewArtist}
          open={showAdd}
        >
          <Artist />
        </FormModal>
      </div>
      <div className="artist-list-container list-container" ref={topRef}>
        <ul className="artist-list card-list">
          {artistCollection?.current == null
            ? ""
            : artistCollection.current.map((artist) => {
                const key = `artist-list-${artist.get("id")}`;
                return <ArtistListItem className key={key} artist={artist} />;
              })}
        </ul>
      </div>
    </div>
  );
}

export default ArtistList;
