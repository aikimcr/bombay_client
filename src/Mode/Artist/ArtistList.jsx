import { useState, useContext, createRef } from 'react';

import './ArtistList.scss';

import BombayLoginContext from '../../Context/BombayLoginContext';

import { ArtistCollection } from '../../Model/ArtistCollection';

import ArtistListItem from './ArtistListItem.jsx';
import Artist from './Artist.jsx';
import FormModal from '../../Modal/FormModal.jsx';
import { useModelCollection } from '../../Hooks/useModelCollection';
import { ProtectedRoute } from '../../Components';

export const ArtistList = (props) => {
  const topRef = createRef();

  const loginState = useContext(BombayLoginContext);

  const [artistCollection] = useState(new ArtistCollection({}));
  const { refreshCollection } = useModelCollection({
    initialCollection: artistCollection,
    topRef,
  });

  const [showAdd, setShowAdd] = useState(false);

  async function submitNewArtist(artistDef) {
    await artistCollection.save(artistDef);
    refreshCollection();
  }

  return (
    <ProtectedRoute>
      <div className="list-component" data-testid="artist-list-component">
        <div className="list-controls">
          <button className="btn" onClick={() => setShowAdd(true)}>
            New
          </button>
          <div className="title">Artists</div>
          <button className="btn" onClick={refreshCollection}>
            Refresh
          </button>
          <FormModal
            title="Add Artist"
            onClose={() => setShowAdd(false)}
            onSubmit={submitNewArtist}
            open={showAdd}
          >
            <Artist />
          </FormModal>
        </div>
        <div className="artist-list-container list-container" ref={topRef}>
          <ul className="artist-list card-list">
            {artistCollection == null
              ? ''
              : artistCollection.map((artist) => {
                  const key = `artist-list-${artist.id}`;
                  return <ArtistListItem className key={key} artist={artist} />;
                })}
          </ul>
        </div>
      </div>
    </ProtectedRoute>
  );
};
