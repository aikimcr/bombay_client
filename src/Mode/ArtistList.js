import { useEffect, useState, useRef, createRef } from 'react';
import useIntersectionObserver from '../Hooks/useIntersectionObserver';

import './ArtistList.scss';

import ArtistCollection from '../Model/ArtistCollection';

import ArtistListItem from './ArtistListItem';
import Artist from './Artist';
import { FormModal } from '../Modal/FormModal';

function ArtistList(props) {
    const topRef = createRef();

    const artistCollection = useRef(null);
    const observer = useIntersectionObserver(topRef, (entries, observer) => {
        entries.forEach(entry => {
            if (entry.isIntersecting) {
                setShouldPage(true);
                observer.unobserve(entry.target);
            }

            // I haven't found a case where this is actually needed, but I haven't ruled it out.
            // if (entry.isVisible) {
            // }
        });
    });

    const [showAdd, setShowAdd] = useState(false);
    const [shouldPage, setShouldPage] = useState(false);
    const [loading, setLoading] = useState(true);

    useEffect(() => {
        const intervalHandle = setInterval(refreshCollection, 300000);

        return () => {
            clearInterval(intervalHandle);
        }
    }, []);

    useEffect(() => {
        if (artistCollection.current == null) {
            refreshCollection()
        }
    }, [artistCollection.current]);

    useEffect(() => {
        if (!shouldPage) { return; }
        if (loading) { return; }

        setShouldPage(false);

        if (artistCollection.current.hasNextPage()) {
            setLoading(true);

            artistCollection.current.fetchNextPage()
                .then(() => {
                    setLoading(false);
                });
        }
    }, [shouldPage]);

    useEffect(() => {
        if (loading) { return; }

        if (artistCollection.current && artistCollection.current.hasNextPage()) {
            const myElement = topRef.current.querySelector('li:last-child');

            if (myElement) {
                observer.current.observe(myElement);
            }
        }
    }, [loading]);

    function refreshCollection() {
        setLoading(true);

        artistCollection.current = new ArtistCollection();
        artistCollection.current.ready()
            .then(() => {
                setLoading(false);
            });
    }

    async function submitNewArtist(artistDef) {
        await artistCollection.current.save(artistDef);
        refreshCollection();
    }

    return (
        <div className="list-component">
            <div className="list-controls">
                <button className="btn" onClick={() => setShowAdd(true)}>New</button>
                <button className="btn" onClick={refreshCollection}>Refresh</button>
                <FormModal
                    title="Add Artist"
                    onClose={() => setShowAdd(false)}
                    onSubmit={submitNewArtist}
                    open={showAdd}>
                    <Artist />
                </FormModal>
            </div>
            <div className="artist-list-container" ref={topRef}>
                <ul className="artist-list card-list">
                    {artistCollection?.current == null ? '' : artistCollection.current.map(artist => {
                        const key = `artist-list-${artist.get('id')}`;
                        return <ArtistListItem className key={key} artist={artist} />
                    })}
                </ul>
            </div>
        </div>
    );
}

export default ArtistList;