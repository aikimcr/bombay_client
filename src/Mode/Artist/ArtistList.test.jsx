import { useState } from "react";
import { BrowserRouter } from "react-router";

import { act, render, screen } from "@testing-library/react";

import {
  mockLoginStatus,
  mockRefreshToken,
  mockLogin,
  mockLogout,
} from "../../Network/testing";

jest.mock("../../Network/Login", () => {
  const originalModule = jest.requireActual("../../Network/Login");

  return {
    __esModule: true,
    ...originalModule,
    loginStatus: mockLoginStatus,
    refreshToken: mockRefreshToken,
    login: mockLogin,
    logout: mockLogout,
  };
});

import { mockModelFetcher, mockUseModelCollection } from "../../Hooks/testing";
jest.mock("../../Hooks/useModelCollection", () => {
  const originalModule = jest.requireActual("../../Hooks/useModelCollection");

  return {
    __esModule: true,
    ...originalModule,
    useModelCollection: mockUseModelCollection,
  };
});

import * as Network from "../../Network/Network";
import * as mockObserver from "../../Hooks/useIntersectionObserver";

import { makeModels, makeAModel } from "../../testHelpers/modelTools";

import BombayLoginContext from "../../Context/BombayLoginContext";
import BombayUtilityContext from "../../Context/BombayUtilityContext";

import { ArtistList } from "./ArtistList.jsx";

jest.useFakeTimers();

function FakeContent(props) {
  const [loginState, setLoginState] = useState({
    loggedIn: true,
    showLoginForm: false,
  });
  const [modeState, setModeState] = useState("artist");

  const checkLoginState = async () => {
    setLoginState(true);
  };

  const setMode = async (newMode) => {
    setModeState("artist");
  };

  const utilities = {
    checkLoginState,
    setMode,
  };

  return (
    <BrowserRouter basename="/">
      <BombayLoginContext.Provider value={loginState}>
        <BombayUtilityContext.Provider value={utilities}>
          {props.children}
        </BombayUtilityContext.Provider>
      </BombayLoginContext.Provider>
    </BrowserRouter>
  );
}

jest.useFakeTimers();

const testToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJGREM4MTEzOCIsInVzZXIiOnsiaWQiOjEsIm5hbWUiOiJhZG1pbiIsImFkbWluIjpmYWxzZX0sImlhdCI6MTY2NTk2NTA5OX0.2vz14X7Tm-oFlyOa7dcAF-5y5ympi_UlWyJNxO4xyS4";

beforeEach(() => {
  localStorage.setItem("jwttoken", testToken);
});

afterEach(() => {
  while (mockObserver.mockObserver.observers.length > 0) {
    mockObserver.mockObserver.observers.pop();
  }

  localStorage.removeItem("jwttoken");
});

function getAreas(result) {
  expect(mockObserver.mockObserver.observers.length).toBe(1);
  const observer = mockObserver.mockObserver.observers[0];

  const index = result.container;
  expect(index.childElementCount).toEqual(1);

  const listComponent = index.firstChild;
  expect(listComponent.childElementCount).toEqual(2);

  const controls = listComponent.firstChild;
  expect(controls.className).toEqual("list-controls");
  expect(controls.childElementCount).toEqual(3);

  const listContainer = listComponent.lastChild;
  expect(listContainer).toHaveClass("artist-list-container");
  expect(listContainer).toHaveClass("list-container");
  expect(listContainer.childElementCount).toEqual(1);
  expect(observer.options.root.lastChild).toHaveClass("artist-list-container");

  const listElement = listContainer.firstChild;

  return [listElement, controls, observer, listContainer, index];
}

const setUpModels = () => {
  const artistNames = [
    "Same & Dave",
    "Aerosmith",
    "Prince",
    "DNCE",
    "Sabrina Carpenter",
    "Metallica",
    "Ben E. King",
    "Van Halen",
    "Led Zeppelin",
    "Black Sabbath",
    "Steppenwolf",
    "Tom Petty",
    "Rolling Stones",
    "Olivia Rodrigo",
    "Santana",
    "The Who",
    "Deep Purple",
    "Donovan",
    "Jimi Hendrix",
    "T. Rex",
  ];
  const [_fetchBody, models] = makeModels(10, {}, "artist", (def) => {
    const idString = def.id || "1";
    const id = parseInt(idString) - 1;
    def.name = artistNames[id];
  });

  const modelPromise = PromiseWithResolvers();
  mockModelFetcher.mockReturnValueOnce(modelPromise.promise);
  return [modelPromise, models];
};

it("Component should match snapshot", async () => {
  const [modelPromise, models] = setUpModels();

  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const { asFragment } = render(
    <FakeContent>
      <ArtistList />
    </FakeContent>,
  );

  await act(async () => {
    modelPromise.resolve(models);
  });

  expect(asFragment()).toMatchSnapshot();
});

it("should render the list", async () => {
  const [modelPromise, models] = setUpModels();

  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const result = render(
    <FakeContent>
      <ArtistList />
    </FakeContent>,
  );

  await act(async () => {
    modelPromise.resolve(models);
  });

  expect(mockObserver.mockObserver.observers.length).toBe(1);

  expect(screen.getByTestId("artist-list-component")).toBeInTheDocument();
  expect(screen.getAllByTestId("artist-list-card")).toHaveLength(models.length);
});

it("should render the next page", async () => {
  const [modelPromise1, models1] = setUpModels();
  const [modelPromise2, models2] = setUpModels();

  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const result = render(
    <FakeContent>
      <ArtistList />
    </FakeContent>,
  );

  await act(async () => {
    modelPromise1.resolve(models1);
  });

  expect(mockObserver.mockObserver.observers.length).toBe(1);
  expect(screen.getByTestId("artist-list-component")).toBeInTheDocument();
  expect(screen.getAllByTestId("artist-list-card")).toHaveLength(
    models1.length,
  );

  const observer = mockObserver.mockObserver.observers[0];

  await act(async () => {
    observer._fireIntersect(
      screen.getAllByTestId("artist-list-card").slice(-1)[0],
    );
    modelPromise2.resolve(models2);
  });

  expect(screen.getAllByTestId("artist-list-card")).toHaveLength(
    models1.length + models2.length,
  );
});

it("should stop when it runs out of data", async () => {
  const [modelPromise, models] = setUpModels();

  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const result = render(
    <FakeContent>
      <ArtistList />
    </FakeContent>,
  );

  await act(async () => {
    modelPromise.resolve(models);
  });

  expect(mockObserver.mockObserver.observers.length).toBe(1);
  expect(screen.getByTestId("artist-list-component")).toBeInTheDocument();
  expect(screen.getAllByTestId("artist-list-card")).toHaveLength(models.length);

  const observer = mockObserver.mockObserver.observers[0];

  await act(async () => {
    observer._fireIntersect(
      screen.getAllByTestId("artist-list-card").slice(-1)[0],
    );
    modelPromise.reject({ status: 404, message: "Not Found" });
  });

  expect(mockObserver.mockObserver.observers.length).toBe(1);
  expect(screen.getByTestId("artist-list-component")).toBeInTheDocument();
  expect(screen.getAllByTestId("artist-list-card")).toHaveLength(models.length);
});

// Move this to testing the artist editing component.
it.skip("should add an artist", async () => {
  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn: true,
    token: testToken,
  });

  const modalRoot = document.getElementById("modal-root");

  const result = render(
    <FakeContent>
      <ArtistList />
    </FakeContent>,
  );

  const [fetchBody] = makeModels(10, {}, "artist");
  const collectionUrl = Network.prepareURLFromArgs("artist").toString();

  await act(async () => {
    resolve(fetchBody);
  });

  expect(mockObserver.mockObserver.observers.length).toBe(1);

  const [, controls] = getAreas(result);

  expect(modalRoot.childElementCount).toEqual(0);

  const addButton = controls.firstChild;

  act(() => {
    addButton.click();
  });

  expect(modalRoot.childElementCount).toEqual(1);

  const submitButton = modalRoot.querySelector('[type="submit"]');

  const [saveDef] = makeAModel("artist");
  await changeInput(
    modalRoot.querySelector('[data-targetfield="name"'),
    "input",
    "Herkimer",
    250,
  );

  await act(async () => {
    submitButton.click();
  });

  resolve(saveDef);

  expect(Network.postToURLString).toBeCalledTimes(1);
  expect(Network.postToURLString).toBeCalledWith(collectionUrl, {
    name: "Herkimer",
  });
});
