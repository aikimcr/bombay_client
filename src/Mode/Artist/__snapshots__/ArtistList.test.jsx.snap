// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="list-component"
    data-testid="artist-list-component"
  >
    <div
      class="list-controls"
    >
      <button
        class="btn"
      >
        New
      </button>
      <div
        class="title"
      >
        Artists
      </div>
      <button
        class="btn"
      >
        Refresh
      </button>
    </div>
    <div
      class="artist-list-container list-container"
    >
      <ul
        class="artist-list card-list"
      >
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            4 Non Blondes
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            AC/DC
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            Aretha Franklin
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            Aerosmith
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            Al Green
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            Alanis Morisette
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            B.B. King
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            Beatles
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            Bee Gees
          </div>
          <div
            class="details"
          />
        </li>
        <li
          class="card"
          data-testid="artist-list-card"
        >
          <div
            class="header"
          >
            Artist
          </div>
          <div
            class="name"
          >
            Ben E. King
          </div>
          <div
            class="details"
          />
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
